feat(starknet_os): implement initialize_state_changes hint

aner-starkware opened the pull request.

reviewable-StarkWare commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/starkware-libs/sequencer/4946)
<!-- Reviewable:end -->

nimrod-starkware reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/4946)* status: 0 of 2 files reviewed, 1 unresolved discussion (waiting on @yoavGrs)

___
*[`crates/starknet_os/src/hints/hint_implementation/os.rs` line 69 at r3](https://reviewable.io/reviews/starkware-libs/sequencer/4946#-OLbyxKz7dLK6Jec5LTZ:-OLbyxL-3TqL8z_ocMYN:bjnr413) ([raw file](https://github.com/starkware-libs/sequencer/blob/a55d5d7f433c93f904d2968ad83db788dcba6f65/crates/starknet_os/src/hints/hint_implementation/os.rs#L69)):*
> ```XML
>     exec_scopes.insert_value(Scope::InitialDict.into(), hashes_and_nonces);
>     Ok(())
> }
> ```

From what I understand, before calling `dict_new` we need to create the hint variable `initial_dict: HashMap<MaybeRelocatable, MaybeRelocatable> `.
So you should be consistent with that.
I do think that you need to call `segments.add()` to allocate space for the storage updates of each contract.
You also need to allocate space for each entry (like moonsong did)



_Code quote:_
```XML
pub(crate) fn initialize_state_changes<S: StateReader>(
    HintArgs { hint_processor, exec_scopes, .. }: HintArgs<'_, S>,
) -> OsHintResult {
    let contract_items = &hint_processor.execution_helper.cached_state.get_initial_reads()?;
    let hashes_and_nonces: Vec<(ClassHash, Nonce)> = contract_items
        .nonces
        .iter()
        .map(|(address, nonce)| {
            let contract_hash = contract_items.class_hashes.get(address).unwrap();
            (*contract_hash, *nonce)
        })
        .collect();
    // TODO(Aner): verify that it is not necessary to recursively iterate over the vector, and
    // insert the values to the vm one by one, or to call segments.add().
    exec_scopes.insert_value(Scope::InitialDict.into(), hashes_and_nonces);
    Ok(())
}
```


<!-- Sent from Reviewable.io -->

aner-starkware commented:
*[`crates/starknet_os/src/hints/hint_implementation/os.rs` line 69 at r3](https://reviewable.io/reviews/starkware-libs/sequencer/4946#-OLbyxKz7dLK6Jec5LTZ:-OLcEivf1jX-Uacxxsn9:bqnuurh) ([raw file](https://github.com/starkware-libs/sequencer/blob/a55d5d7f433c93f904d2968ad83db788dcba6f65/crates/starknet_os/src/hints/hint_implementation/os.rs#L69)):*
<details><summary><i>Previously, nimrod-starkware wrote…</i></summary><blockquote>

From what I understand, before calling `dict_new` we need to create the hint variable `initial_dict: HashMap<MaybeRelocatable, MaybeRelocatable> `.
So you should be consistent with that.
I do think that you need to call `segments.add()` to allocate space for the storage updates of each contract.
You also need to allocate space for each entry (like moonsong did)
</blockquote></details>

You're not looking at the latest version. The question is if this isn't what `vm.gen_arg` already does.


<!-- Sent from Reviewable.io -->

nimrod-starkware reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/4946)* status: 0 of 2 files reviewed, 1 unresolved discussion (waiting on @yoavGrs)

___
*[`crates/starknet_os/src/hints/hint_implementation/os.rs` line 39 at r4](https://reviewable.io/reviews/starkware-libs/sequencer/4946#-OLcK_VlE-oa8mHI9P0X:-OLcK_VlE-oa8mHI9P0Y:b-h23y4s) ([raw file](https://github.com/starkware-libs/sequencer/blob/05696979b1ff4833cecb2352dad3a2ce11130b3a/crates/starknet_os/src/hints/hint_implementation/os.rs#L39)):*
> ```XML
>     exec_scopes.insert_value(Scope::InitialDict.into(), initial_dict);
>     Ok(())
> }
> ```

by looking [here](https://github.com/lambdaclass/cairo-vm/blob/main/vm/src/vm/vm_memory/memory_segments.rs#L132) , it looks like `gen_arg` can't take `Vec<(MaybeRelocatable, MaybeRelocatable)>`. 
And you should allocate space for the field `storage_ptr` of `StateEntry` (with `add_segment` or other method)



_Code quote:_
```XML
    let contract_items = &hint_processor.execution_helper.cached_state.get_initial_reads()?;
    let hashes_and_nonces: Vec<(ClassHash, Nonce)> = contract_items
        .nonces
        .iter()
        .map(|(address, nonce)| {
            let contract_hash =
                contract_items.class_hashes.get(address).expect("Missing contract hash");
            (*contract_hash, *nonce)
        })
        .collect();
    // TODO(Aner): verify that it is not necessary to add segments.add().
    let initial_dict = vm.gen_arg(&hashes_and_nonces)?;
    exec_scopes.insert_value(Scope::InitialDict.into(), initial_dict);
    Ok(())
}
```


<!-- Sent from Reviewable.io -->

nimrod-starkware reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/4946)* status: 0 of 2 files reviewed, 3 unresolved discussions (waiting on @aner-starkware and @yoavGrs)

___
*[`crates/blockifier/src/state/cached_state.rs` line 106 at r4](https://reviewable.io/reviews/starkware-libs/sequencer/4946#-OLcLSjuDbjkxbUjFlS6:-OLcLSjuDbjkxbUjFlS7:bfmdpts) ([raw file](https://github.com/starkware-libs/sequencer/blob/05696979b1ff4833cecb2352dad3a2ce11130b3a/crates/blockifier/src/state/cached_state.rs#L106)):*
> ```XML
>     pub fn get_initial_reads(&self) -> StateResult<StateMaps> {
>         Ok(self.cache.borrow().initial_reads.clone())
>     }
> ```

Why is that needed?



_Code quote:_
```XML
    pub fn get_initial_reads(&self) -> StateResult<StateMaps> {
        Ok(self.cache.borrow().initial_reads.clone())
    }
```
___
*[`crates/starknet_os/src/hints/hint_implementation/os.rs` line 25 at r4](https://reviewable.io/reviews/starkware-libs/sequencer/4946#-OLcLWVg7VOu0HwpkDjT:-OLcLWVg7VOu0HwpkDjU:b-qzr7ge) ([raw file](https://github.com/starkware-libs/sequencer/blob/05696979b1ff4833cecb2352dad3a2ce11130b3a/crates/starknet_os/src/hints/hint_implementation/os.rs#L25)):*
> ```XML
>     HintArgs { hint_processor, exec_scopes, vm, .. }: HintArgs<'_, S>,
> ) -> OsHintResult {
>     let contract_items = &hint_processor.execution_helper.cached_state.get_initial_reads()?;
> ```

Why not use the methods `get_nonce_at` and ` get_class_hash_at`? 



_Code quote:_
```XML
let contract_items = &hint_processor.execution_helper.cached_state.get_initial_reads()?;
```


<!-- Sent from Reviewable.io -->

aner-starkware commented:
*[`crates/starknet_os/src/hints/hint_implementation/os.rs` line 25 at r4](https://reviewable.io/reviews/starkware-libs/sequencer/4946#-OLcLWVg7VOu0HwpkDjT:-OLcO0DSCG4KPBrtS4wo:bewm40s) ([raw file](https://github.com/starkware-libs/sequencer/blob/05696979b1ff4833cecb2352dad3a2ce11130b3a/crates/starknet_os/src/hints/hint_implementation/os.rs#L25)):*
<details><summary><i>Previously, nimrod-starkware wrote…</i></summary><blockquote>

Why not use the methods `get_nonce_at` and ` get_class_hash_at`?
</blockquote></details>

For which addresses?


<!-- Sent from Reviewable.io -->

yoavGrs reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/4946)* status: 0 of 2 files reviewed, 5 unresolved discussions (waiting on @aner-starkware and @nimrod-starkware)

___
*[`crates/starknet_os/src/hints/hint_implementation/os.rs` line 32 at r4](https://reviewable.io/reviews/starkware-libs/sequencer/4946#-OLcYJzq8-NjZuRatx0m:-OLcYJzq8-NjZuRatx0n:b-iira9q) ([raw file](https://github.com/starkware-libs/sequencer/blob/05696979b1ff4833cecb2352dad3a2ce11130b3a/crates/starknet_os/src/hints/hint_implementation/os.rs#L32)):*
> ```XML
>             let contract_hash =
>                 contract_items.class_hashes.get(address).expect("Missing contract hash");
>             (*contract_hash, *nonce)
> ```

It should be a mapping from a contract address to a triple.



_Code quote:_
```XML
(*contract_hash, *nonce)
```


<!-- Sent from Reviewable.io -->

aner-starkware reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/4946)* status: 0 of 3 files reviewed, 5 unresolved discussions (waiting on @nimrod-starkware and @yoavGrs)

___
*[`crates/starknet_os/src/hints/hint_implementation/os.rs` line 25 at r4](https://reviewable.io/reviews/starkware-libs/sequencer/4946#-OLcX2l22XaLebTkwV37:-OLcaxus0c-JOkK4vkcy:b-4hp6ax) ([raw file](https://github.com/starkware-libs/sequencer/blob/05696979b1ff4833cecb2352dad3a2ce11130b3a/crates/starknet_os/src/hints/hint_implementation/os.rs#L25)):*
<details><summary><i>Previously, yoavGrs wrote…</i></summary><blockquote>

Are you sure about it?
I see in the Cairo code this dict is going into `contract_state_changes` variable.
</blockquote></details>

No, I'm not sure. The question is if the OS input (specifically, the `contracts`) is modified during the OS run. If it is, then it should be switched from the `initial_reads` `StateMaps` to the `writes` `StateMaps`. I'll try to check.
___
*[`crates/starknet_os/src/hints/hint_implementation/os.rs` line 32 at r4](https://reviewable.io/reviews/starkware-libs/sequencer/4946#-OLcYJzq8-NjZuRatx0m:-OLcavjh8lIHpBLiB0RN:b-896fix) ([raw file](https://github.com/starkware-libs/sequencer/blob/05696979b1ff4833cecb2352dad3a2ce11130b3a/crates/starknet_os/src/hints/hint_implementation/os.rs#L32)):*
<details><summary><i>Previously, yoavGrs wrote…</i></summary><blockquote>

It should be a mapping from a contract address to a triple.
</blockquote></details>

Done.
___
*[`crates/starknet_os/src/hints/hint_implementation/os.rs` line 39 at r4](https://reviewable.io/reviews/starkware-libs/sequencer/4946#-OLcK_VlE-oa8mHI9P0X:-OLcaweR-_WOkxfV236V:b-896fix) ([raw file](https://github.com/starkware-libs/sequencer/blob/05696979b1ff4833cecb2352dad3a2ce11130b3a/crates/starknet_os/src/hints/hint_implementation/os.rs#L39)):*
<details><summary><i>Previously, nimrod-starkware wrote…</i></summary><blockquote>

by looking [here](https://github.com/lambdaclass/cairo-vm/blob/main/vm/src/vm/vm_memory/memory_segments.rs#L132) , it looks like `gen_arg` can't take `Vec<(MaybeRelocatable, MaybeRelocatable)>`. 
And you should allocate space for the field `storage_ptr` of `StateEntry` (with `add_segment` or other method)
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

aner-starkware commented:
*[`crates/starknet_os/src/hints/hint_implementation/os.rs` line 25 at r4](https://reviewable.io/reviews/starkware-libs/sequencer/4946#-OLcX2l22XaLebTkwV37:-OLccW8w8Re7Ya0lCzpH:b-wrkkz4) ([raw file](https://github.com/starkware-libs/sequencer/blob/05696979b1ff4833cecb2352dad3a2ce11130b3a/crates/starknet_os/src/hints/hint_implementation/os.rs#L25)):*
<details><summary><i>Previously, aner-starkware wrote…</i></summary><blockquote>

No, I'm not sure. The question is if the OS input (specifically, the `contracts`) is modified during the OS run. If it is, then it should be switched from the `initial_reads` `StateMaps` to the `writes` `StateMaps`. I'll try to check.
</blockquote></details>

Also, the function is called `initialize_state_changes`, which suggests that `initial_reads` is the right one.


<!-- Sent from Reviewable.io -->

aner-starkware reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/4946)* status: 0 of 4 files reviewed, 7 unresolved discussions (waiting on @nimrod-starkware and @yoavGrs)

___
*[`crates/starknet_os/src/hints/hint_implementation/os.rs` line 25 at r4](https://reviewable.io/reviews/starkware-libs/sequencer/4946#-OLcX2l22XaLebTkwV37:-OLdDZu71TYVtnzUOT22:b45y6ay) ([raw file](https://github.com/starkware-libs/sequencer/blob/05696979b1ff4833cecb2352dad3a2ce11130b3a/crates/starknet_os/src/hints/hint_implementation/os.rs#L25)):*
<details><summary><i>Previously, aner-starkware wrote…</i></summary><blockquote>

Also, the function is called `initialize_state_changes`, which suggests that `initial_reads` is the right one.
</blockquote></details>

@nimrod-starkware  and I looked at it - it seems that `initial_reads` is not used in the OS, so I changed it to `writes`.
___
*[`crates/starknet_os/src/hints/hint_implementation/os.rs` line 36 at r5](https://reviewable.io/reviews/starkware-libs/sequencer/4946#-OLchQFIAzXnW3BLx7ay:-OLdDYXW3eGRcav_mjou:b-896fix) ([raw file](https://github.com/starkware-libs/sequencer/blob/405e29be7beb631d008ccbf9315e7b19739534db/crates/starknet_os/src/hints/hint_implementation/os.rs#L36)):*
<details><summary><i>Previously, nimrod-starkware wrote…</i></summary><blockquote>

please use `insert_values_to_fields` for this - it allows you to remove the hard coded offsets
</blockquote></details>

Done.
___
*[`crates/starknet_os/src/hints/hint_implementation/os.rs` line 28 at r6](https://reviewable.io/reviews/starkware-libs/sequencer/4946#-OLdDvOi4-O5mJIHh8gn:-OLdDvOi4-O5mJIHh8go:b-72f0qx) ([raw file](https://github.com/starkware-libs/sequencer/blob/271dc1b85f186b35d84bdecf4090eaea0a405ade/crates/starknet_os/src/hints/hint_implementation/os.rs#L28)):*
> ```XML
> ) -> OsHintResult {
>     let state_maps = get_state_maps(hint_processor)?;
>     let class_hash_to_compiled_class_hash = state_maps.compiled_class_hashes;
> ```

@nimrod-starkware , according to what we looked at this morning, I think this should probably also be changed to `writes`



_Code quote:_
```XML
    let state_maps = get_state_maps(hint_processor)?;
    let class_hash_to_compiled_class_hash = state_maps.compiled_class_hashes;
```


<!-- Sent from Reviewable.io -->

aner-starkware commented:
*[`crates/blockifier/src/state/cached_state.rs` line 106 at r4](https://reviewable.io/reviews/starkware-libs/sequencer/4946#-OLcLSjuDbjkxbUjFlS6:-OLdEDZr06FlIFK3jLKM:bdti5j2) ([raw file](https://github.com/starkware-libs/sequencer/blob/05696979b1ff4833cecb2352dad3a2ce11130b3a/crates/blockifier/src/state/cached_state.rs#L106)):*
<details><summary><i>Previously, aner-starkware wrote…</i></summary><blockquote>

The `StarknetOsInput` field `contracts: Dict[int, ContractState]` isn't passed to the Rust OS. IIUC, this is the way to get the required values.
</blockquote></details>

Change to `writes`, according to what we looked at this morning.


<!-- Sent from Reviewable.io -->

nimrod-starkware reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/4946)* status: 0 of 4 files reviewed, 8 unresolved discussions (waiting on @aner-starkware and @yoavGrs)

___
*[`crates/starknet_os/src/hints/hint_implementation/os.rs` line 39 at r6](https://reviewable.io/reviews/starkware-libs/sequencer/4946#-OLdFT5A0pKm08tFU2Dv:-OLdFT5B7VHxrIEVIgEy:b-7xlsrd) ([raw file](https://github.com/starkware-libs/sequencer/blob/271dc1b85f186b35d84bdecf4090eaea0a405ade/crates/starknet_os/src/hints/hint_implementation/os.rs#L39)):*
> ```XML
>     let mut initial_dict: HashMap<MaybeRelocatable, MaybeRelocatable> = HashMap::new();
>     for (address, nonce) in state_maps.nonces.iter() {
>         let contract_hash = state_maps.class_hashes.get(address).expect("Missing contract hash");
> ```

rename please



_Suggestion:_
```XML
class_hash
```
___
*[`crates/starknet_os/src/hints/hint_implementation/os.rs` line 40 at r6](https://reviewable.io/reviews/starkware-libs/sequencer/4946#-OLdFWKa7DfNTeujfW03:-OLdFWKb7mXDgFFmUgGS:b-7xlsrd) ([raw file](https://github.com/starkware-libs/sequencer/blob/271dc1b85f186b35d84bdecf4090eaea0a405ade/crates/starknet_os/src/hints/hint_implementation/os.rs#L40)):*
> ```XML
>     for (address, nonce) in state_maps.nonces.iter() {
>         let contract_hash = state_maps.class_hashes.get(address).expect("Missing contract hash");
>         let change_base = vm.add_memory_segment();
> ```

rename please



_Suggestion:_
```XML
state_entry_base
```
___
*[`crates/starknet_os/src/hints/hint_implementation/os.rs` line 41 at r6](https://reviewable.io/reviews/starkware-libs/sequencer/4946#-OLdFMjQ5H7odvup12nJ:-OLdFMjQ5H7odvup12nK:b-l6mkjh) ([raw file](https://github.com/starkware-libs/sequencer/blob/271dc1b85f186b35d84bdecf4090eaea0a405ade/crates/starknet_os/src/hints/hint_implementation/os.rs#L41)):*
> ```XML
>         let contract_hash = state_maps.class_hashes.get(address).expect("Missing contract hash");
>         let change_base = vm.add_memory_segment();
>         let new_segment = vm.add_memory_segment();
> ```

rename



_Suggestion:_
```XML
storage_ptr
```
___
*[`crates/starknet_os/src/hints/hint_implementation/os.rs` line 48 at r6](https://reviewable.io/reviews/starkware-libs/sequencer/4946#-OLdFJ4l7VtVVtkPPoh5:-OLdFJ4l7VtVVtkPPoh6:undefined) ([raw file](https://github.com/starkware-libs/sequencer/blob/271dc1b85f186b35d84bdecf4090eaea0a405ade/crates/starknet_os/src/hints/hint_implementation/os.rs#L48)):*
> ```XML
>             &[
>                 ("class_hash".to_string(), MaybeRelocatable::from(**contract_hash)),
>                 ("new_segment".to_string(), new_segment.into()),
> ```





_Suggestion:_
```XML
("storage_ptr".to_string(), new_segment.into())
```


<!-- Sent from Reviewable.io -->

aner-starkware reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/4946)* status: 0 of 4 files reviewed, 8 unresolved discussions (waiting on @nimrod-starkware and @yoavGrs)

___
*[`crates/starknet_os/src/hints/hint_implementation/os.rs` line 25 at r4](https://reviewable.io/reviews/starkware-libs/sequencer/4946#-OLcLWVg7VOu0HwpkDjT:-OLdP4IADveoVeHAYN5I:b-u5v4ym) ([raw file](https://github.com/starkware-libs/sequencer/blob/05696979b1ff4833cecb2352dad3a2ce11130b3a/crates/starknet_os/src/hints/hint_implementation/os.rs#L25)):*
<details><summary><i>Previously, aner-starkware wrote…</i></summary><blockquote>

For which addresses?
</blockquote></details>

See here:
https://reviewable.io/reviews/starkware-libs/sequencer/5023
___
*[`crates/starknet_os/src/hints/hint_implementation/os.rs` line 39 at r6](https://reviewable.io/reviews/starkware-libs/sequencer/4946#-OLdFT5A0pKm08tFU2Dv:-OLdQs-7CiWS_ADeMjPy:b-896fix) ([raw file](https://github.com/starkware-libs/sequencer/blob/271dc1b85f186b35d84bdecf4090eaea0a405ade/crates/starknet_os/src/hints/hint_implementation/os.rs#L39)):*
<details><summary><i>Previously, nimrod-starkware wrote…</i></summary><blockquote>

rename please
</blockquote></details>

Done.
___
*[`crates/starknet_os/src/hints/hint_implementation/os.rs` line 40 at r6](https://reviewable.io/reviews/starkware-libs/sequencer/4946#-OLdFWKa7DfNTeujfW03:-OLdQsDsDfQ9t8cqzFR3:b-896fix) ([raw file](https://github.com/starkware-libs/sequencer/blob/271dc1b85f186b35d84bdecf4090eaea0a405ade/crates/starknet_os/src/hints/hint_implementation/os.rs#L40)):*
<details><summary><i>Previously, nimrod-starkware wrote…</i></summary><blockquote>

rename please
</blockquote></details>

Done.
___
*[`crates/starknet_os/src/hints/hint_implementation/os.rs` line 41 at r6](https://reviewable.io/reviews/starkware-libs/sequencer/4946#-OLdFMjQ5H7odvup12nJ:-OLdQsas-NZ1FFuzN2Rm:b-896fix) ([raw file](https://github.com/starkware-libs/sequencer/blob/271dc1b85f186b35d84bdecf4090eaea0a405ade/crates/starknet_os/src/hints/hint_implementation/os.rs#L41)):*
<details><summary><i>Previously, nimrod-starkware wrote…</i></summary><blockquote>

rename
</blockquote></details>

Done.
___
*[`crates/starknet_os/src/hints/hint_implementation/os.rs` line 48 at r6](https://reviewable.io/reviews/starkware-libs/sequencer/4946#-OLdFJ4l7VtVVtkPPoh5:-OLdQtY08gGTPC1jk-9B:b-896fix) ([raw file](https://github.com/starkware-libs/sequencer/blob/271dc1b85f186b35d84bdecf4090eaea0a405ade/crates/starknet_os/src/hints/hint_implementation/os.rs#L48)):*
> ```XML
>             &[
>                 ("class_hash".to_string(), MaybeRelocatable::from(**contract_hash)),
>                 ("new_segment".to_string(), new_segment.into()),
> ```

Done.


<!-- Sent from Reviewable.io -->

aner-starkware reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/4946)* status: 0 of 4 files reviewed, 11 unresolved discussions (waiting on @nimrod-starkware and @yoavGrs)

___
*[`crates/blockifier/src/state/cached_state.rs` line 381 at r7](https://reviewable.io/reviews/starkware-libs/sequencer/4946#-OLdfO8s7g93VRiCDppa:-OLhpKgt210iZFVDaHgn:bgr3xx0) ([raw file](https://github.com/starkware-libs/sequencer/blob/5a7c2924f3c1f9b615b082ba38267215f1ad0202/crates/blockifier/src/state/cached_state.rs#L381)):*
<details><summary><i>Previously, yoavGrs wrote…</i></summary><blockquote>

Not needed now, right?
</blockquote></details>

This makes it more restrictive - before, I removed it; now that it's not needed, I reinstated it.
___
*[`crates/starknet_api/src/core.rs` line 350 at r6](https://reviewable.io/reviews/starkware-libs/sequencer/4946#-OLdllaKDWpwDtGVDiqH:-OLhr6bt43zpwJ20Rpk2:buww8o5) ([raw file](https://github.com/starkware-libs/sequencer/blob/271dc1b85f186b35d84bdecf4090eaea0a405ade/crates/starknet_api/src/core.rs#L350)):*
<details><summary><i>Previously, aner-starkware wrote…</i></summary><blockquote>

Some change is needed somewhere. We can talk about the possibilities tomorrow.
</blockquote></details>

Changed to a `From`.


<!-- Sent from Reviewable.io -->

nimrod-starkware reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/4946)* status: 0 of 4 files reviewed, 12 unresolved discussions (waiting on @aner-starkware and @yoavGrs)

___
*[`crates/starknet_api/src/core.rs` line 393 at r8](https://reviewable.io/reviews/starkware-libs/sequencer/4946#-OLiBOGz-tCnyxa1SR0t:-OLiBOGz-tCnyxa1SR0u:b-ycqupm) ([raw file](https://github.com/starkware-libs/sequencer/blob/7229a967739752d4edfc8e96e4197da5bd7f2823/crates/starknet_api/src/core.rs#L393)):*
> ```XML
>         key.0
>     }
> }
> ```

please remove and use the method `key` to convert it to a felt



_Code quote:_
```XML
impl From<PatriciaKey> for StarkHash {
    fn from(key: PatriciaKey) -> Self {
        key.0
    }
}
```


<!-- Sent from Reviewable.io -->

nimrod-starkware reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/4946)* status: 0 of 4 files reviewed, 9 unresolved discussions (waiting on @aner-starkware and @yoavGrs)

___
*[`crates/starknet_os/src/hints/hint_implementation/os.rs` line 54 at r8](https://reviewable.io/reviews/starkware-libs/sequencer/4946#-OLiC-FE1T6rzq2l_xHS:-OLiC-FE1T6rzq2l_xHT:undefined) ([raw file](https://github.com/starkware-libs/sequencer/blob/7229a967739752d4edfc8e96e4197da5bd7f2823/crates/starknet_os/src/hints/hint_implementation/os.rs#L54)):*
> ```XML
>             &hint_processor.execution_helper.os_program,
>         )?;
>         initial_dict.insert(StarkHash::from(address.0).into(), state_entry_base.into());
> ```





_Suggestion:_
```XML
ley key = *address.0.key();
initial_dict.insert(key.into(), state_entry_base.into());
```


<!-- Sent from Reviewable.io -->

aner-starkware reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/4946)* status: 0 of 4 files reviewed, 9 unresolved discussions (waiting on @nimrod-starkware and @yoavGrs)

___
*[`crates/starknet_api/src/core.rs` line 393 at r8](https://reviewable.io/reviews/starkware-libs/sequencer/4946#-OLiBOGz-tCnyxa1SR0t:-OLiWYXH6agAZ2LivpU4:b-896fix) ([raw file](https://github.com/starkware-libs/sequencer/blob/7229a967739752d4edfc8e96e4197da5bd7f2823/crates/starknet_api/src/core.rs#L393)):*
<details><summary><i>Previously, nimrod-starkware wrote…</i></summary><blockquote>

please remove and use the method `key` to convert it to a felt
</blockquote></details>

Done.
___
*[`crates/starknet_os/src/hints/hint_implementation/os.rs` line 54 at r8](https://reviewable.io/reviews/starkware-libs/sequencer/4946#-OLiC-FE1T6rzq2l_xHS:-OLiWMVu4rf8xTXaDORh:b-wxlzak) ([raw file](https://github.com/starkware-libs/sequencer/blob/7229a967739752d4edfc8e96e4197da5bd7f2823/crates/starknet_os/src/hints/hint_implementation/os.rs#L54)):*
> ```XML
>             &hint_processor.execution_helper.os_program,
>         )?;
>         initial_dict.insert(StarkHash::from(address.0).into(), state_entry_base.into());
> ```

Done. Thanks, missed it.


<!-- Sent from Reviewable.io -->

aner-starkware commented:
*[`crates/starknet_api/src/core.rs` line 350 at r6](https://reviewable.io/reviews/starkware-libs/sequencer/4946#-OLdllaKDWpwDtGVDiqH:-OLihhUg76Q5dv8y1ANI:bdv0xzn) ([raw file](https://github.com/starkware-libs/sequencer/blob/271dc1b85f186b35d84bdecf4090eaea0a405ade/crates/starknet_api/src/core.rs#L350)):*
<details><summary><i>Previously, aner-starkware wrote…</i></summary><blockquote>

Changed to a `From`.
</blockquote></details>

Changed to using the `.key()` fn.


<!-- Sent from Reviewable.io -->

nimrod-starkware reviewed (changes requested):
> Reviewed 1 of 3 files at r6, 1 of 2 files at r10.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/4946)* status: 2 of 4 files reviewed, 9 unresolved discussions (waiting on @aner-starkware and @yoavGrs)

___
*[`crates/starknet_os/src/hints/hint_implementation/os.rs` line 48 at r10](https://reviewable.io/reviews/starkware-libs/sequencer/4946#-OLipNKaE_E3GW0ihtxC:-OLipNKaE_E3GW0ihtxD:b-81ri4o) ([raw file](https://github.com/starkware-libs/sequencer/blob/4375434d01c7fedaa5de41e3b4813365040b661b/crates/starknet_os/src/hints/hint_implementation/os.rs#L48)):*
> ```XML
>             &[
>                 ("class_hash".to_string(), MaybeRelocatable::from(**class_hash)),
>                 ("storage_ptr".to_string(), storage_ptr.into()),
> ```

please be consistent with `into`



_Suggestion:_
```XML
                ("class_hash".to_string(), (**class_hash).into()),
                ("storage_ptr".to_string(), storage_ptr.into()),
```
___
*[`crates/starknet_os/src/hints/hint_implementation/os.rs` line 49 at r10](https://reviewable.io/reviews/starkware-libs/sequencer/4946#-OLipVQg4SacPJGCpyvA:-OLipVQg4SacPJGCpyvB:undefined) ([raw file](https://github.com/starkware-libs/sequencer/blob/4375434d01c7fedaa5de41e3b4813365040b661b/crates/starknet_os/src/hints/hint_implementation/os.rs#L49)):*
> ```XML
>                 ("class_hash".to_string(), MaybeRelocatable::from(**class_hash)),
>                 ("storage_ptr".to_string(), storage_ptr.into()),
>                 ("nonce".to_string(), MaybeRelocatable::from(**nonce)),
> ```





_Suggestion:_
```XML
(**nonce).into()
```


<!-- Sent from Reviewable.io -->

aner-starkware reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/4946)* status: 2 of 4 files reviewed, 9 unresolved discussions (waiting on @aner-starkware, @nimrod-starkware, and @yoavGrs)

___
*[`crates/starknet_os/src/hints/hint_implementation/os.rs` line 48 at r10](https://reviewable.io/reviews/starkware-libs/sequencer/4946#-OLipNKaE_E3GW0ihtxC:-OLmEMka3PGqrXzJr8Bd:b24w24p) ([raw file](https://github.com/starkware-libs/sequencer/blob/4375434d01c7fedaa5de41e3b4813365040b661b/crates/starknet_os/src/hints/hint_implementation/os.rs#L48)):*
<details><summary><i>Previously, nimrod-starkware wrote…</i></summary><blockquote>

please be consistent with `into`
</blockquote></details>

Doesn't work here, the compiler can't manage to infer the type.  But it's resolved in the PRs above it.
___
*[`crates/starknet_os/src/hints/hint_implementation/os.rs` line 49 at r10](https://reviewable.io/reviews/starkware-libs/sequencer/4946#-OLipVQg4SacPJGCpyvA:-OLmERjtCOJHH4cJ8Mzf:b-brncwv) ([raw file](https://github.com/starkware-libs/sequencer/blob/4375434d01c7fedaa5de41e3b4813365040b661b/crates/starknet_os/src/hints/hint_implementation/os.rs#L49)):*
> ```XML
>                 ("class_hash".to_string(), MaybeRelocatable::from(**class_hash)),
>                 ("storage_ptr".to_string(), storage_ptr.into()),
>                 ("nonce".to_string(), MaybeRelocatable::from(**nonce)),
> ```

Doesn't work here, the compiler can't manage to infer the type. But it's resolved in the PRs above it.


<!-- Sent from Reviewable.io -->

yoavGrs reviewed (approved):
> Reviewed 1 of 3 files at r6, 1 of 2 files at r10, 1 of 2 files at r12, 1 of 1 files at r13, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/4946)* status: all files reviewed, 4 unresolved discussions (waiting on @nimrod-starkware)

___
*[`crates/blockifier/src/state/cached_state.rs` line 356 at r13](https://reviewable.io/reviews/starkware-libs/sequencer/4946#-OLmvFob6MhZ-vaxWjH0:-OLmvFoc8ArwOVNeCbfT:bxmpgrl) ([raw file](https://github.com/starkware-libs/sequencer/blob/938935e9c8ec1fd3acb04f567f7a648c2225997d/crates/blockifier/src/state/cached_state.rs#L356)):*
> ```XML
>     }
> 
>     pub fn get_compiled_class_hashes(&self) -> &HashMap<ClassHash, CompiledClassHash> {
> ```

IMO we don't need a getter for a public field.



_Code quote:_
```XML
pub fn get_compiled_class_hashes
```


<!-- Sent from Reviewable.io -->

aner-starkware commented:
*[`crates/blockifier/src/state/cached_state.rs` line 356 at r13](https://reviewable.io/reviews/starkware-libs/sequencer/4946#-OLmvFob6MhZ-vaxWjH0:-OLnICcH6jExWnPDePBl:b-hych7u) ([raw file](https://github.com/starkware-libs/sequencer/blob/938935e9c8ec1fd3acb04f567f7a648c2225997d/crates/blockifier/src/state/cached_state.rs#L356)):*
<details><summary><i>Previously, yoavGrs wrote…</i></summary><blockquote>

IMO we don't need a getter for a public field.
</blockquote></details>

Right; fixed.


<!-- Sent from Reviewable.io -->

nimrod-starkware reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/4946)* status: 3 of 4 files reviewed, 1 unresolved discussion (waiting on @yoavGrs)

___
*[`crates/blockifier/src/state/cached_state.rs` line 356 at r13](https://reviewable.io/reviews/starkware-libs/sequencer/4946#-OLmvFob6MhZ-vaxWjH0:-OLnIEOwBQfZmdUTT1HD:b-cstnkg) ([raw file](https://github.com/starkware-libs/sequencer/blob/938935e9c8ec1fd3acb04f567f7a648c2225997d/crates/blockifier/src/state/cached_state.rs#L356)):*
<details><summary><i>Previously, yoavGrs wrote…</i></summary><blockquote>

IMO we don't need a getter for a public field.
</blockquote></details>

I agree


<!-- Sent from Reviewable.io -->

nimrod-starkware reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/4946#-:-OLnSk_t97Bx-DyC97Ue:bnfp4nl)* status: 3 of 4 files reviewed, all discussions resolved (waiting on @yoavGrs)



<!-- Sent from Reviewable.io -->

yoavGrs reviewed (approved):
> Reviewed 1 of 1 files at r14, 2 of 2 files at r15, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/4946)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved (waiting on @aner-starkware)



<!-- Sent from Reviewable.io -->
